internal/service/cloud/image.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2024 IONOS Cloud.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	sdk "github.com/ionos-cloud/sdk-go/v6"
+	"k8s.io/apimachinery/pkg/labels"
+
+	infrav1 "github.com/ionos-cloud/cluster-api-provider-ionoscloud/api/v1alpha1"
+	"github.com/ionos-cloud/cluster-api-provider-ionoscloud/internal/util/ptr"
+	"github.com/ionos-cloud/cluster-api-provider-ionoscloud/scope"
+)
+
+var errMissingMachineVersion = errors.New("machine is missing version field")
+
+// imageMatchError indicates that either 0 or more than 1 images matched a selector.
+// This error signals that user action is required.
+type imageMatchError struct {
+	imageIDs []string
+	selector *infrav1.ImageSelector
+}
+
+func (e imageMatchError) Error() string {
+	return fmt.Sprintf("found %d images matching selector %q",
+		len(e.imageIDs), labels.SelectorFromSet(e.selector.MatchLabels))
+}
+
+func (s *Service) lookupImageID(ctx context.Context, ms *scope.Machine) (string, error) {
+	imageSpec := ms.IonosMachine.Spec.Disk.Image
+
+	if imageSpec.ID != "" {
+		return imageSpec.ID, nil
+	}
+
+	images, err := s.lookupImagesBySelector(ctx, ms.ClusterScope.Location(), imageSpec.Selector)
```

Comment:
The location is currently being reworked: https://github.com/ionos-cloud/cluster-api-provider-ionoscloud/pull/172.
So we either wait for it to be merged or we need a follow up here

Reply:
#172 merged